src/iso9660.rs

```diff
@@ -0,0 +1,517 @@
+    }
+
+    let address = (eat(buf, 1).get_u32_le() as u64) * (ISO9660_SECTOR_SIZE as u64);
+    let length = eat(buf, 4).get_u32_le();
+    let flags = eat(buf, 25 - 14).get_u8();
+    let name_length = eat(buf, 32 - 26).get_u8() as usize;
+    let name = parse_iso9660_path(buf, name_length).context("parsing record name")?;
+
+    // advance to next record
+    eat(buf, len - (33 + name_length));
+
+    if flags & 2 > 0 {
+        Ok(Some(DirectoryRecord::Directory(Directory {
+            name,
+            address,
+            length,
+        })))
+    } else {
+        Ok(Some(DirectoryRecord::File(File {
+            name,
+            address,
+            length,
+        })))
+    }
+}
+
+/// Reads a directory record path. This is similar to a regular ISO9660 string, but supports '\0'
+/// to mean current directory, and '\1' for the parent directory.
+fn parse_iso9660_path(buf: &mut Bytes, len: usize) -> Result<String> {
+    if len == 1 && (buf[0] == 0 || buf[0] == 1) {
+        let c = buf.get_u8();
+        if c == 0 {
+            Ok(".".into())
+        } else {
+            Ok("..".into())
+        }
+    } else {
+        parse_iso9660_string(buf, len, IsoString::File)
+    }
+}
+
+#[allow(unused)]
+enum IsoString {
+    StrA,
+    StrD,
+    File,
+}
+
+/// Reads an ISO9660 string.
+fn parse_iso9660_string(buf: &mut Bytes, len: usize, kind: IsoString) -> Result<String> {
+    const FILE_CHARS: [u8; 17] = *b"!\"%&'()*+,-.:<=>?";
+    const A_CHARS: [u8; 2] = *b";/";
+    if len > buf.remaining() {
+        bail!("incomplete string name; corrupt ISO?");
+    }
+    let mut s = String::with_capacity(len);
+    let mut bytes = buf.copy_to_bytes(len);
+    if matches!(kind, IsoString::File) {
+        if bytes.ends_with(b";1") {
+            bytes.truncate(bytes.len() - 2);
+        }
+        if bytes.ends_with(b".") {
+            bytes.truncate(bytes.len() - 1);
+        }
+    }
+    for byte in &bytes {
+        #[allow(clippy::if_same_then_else)] // I find it easier to follow this way
+        if byte.is_ascii_alphabetic() || byte.is_ascii_digit() || *byte == b'_' || *byte == b' ' {
+            s.push(char::from(*byte));
+        } else if FILE_CHARS.contains(byte) && matches!(kind, IsoString::File | IsoString::StrA) {
+            s.push(char::from(*byte));
+        } else if A_CHARS.contains(byte) && matches!(kind, IsoString::StrA) {
+            s.push(char::from(*byte));
+        } else if A_CHARS.contains(byte) && matches!(kind, IsoString::File) {
+            s.push('.'); // this matches what the kernel does
+        } else if *byte == 0 {
+            break;
+        } else {
+            bail!("invalid string name {:?}", bytes);
+        }
+    }
+    if matches!(kind, IsoString::StrA | IsoString::StrD) {
+        s.truncate(s.trim_end_matches(' ').len());
+    } else {
+        // IsoString::File -- let's make lowercase to be less screamy
+        s.make_ascii_lowercase();
```

Comment:
Hmm, this is what the kernel itself does though: https://github.com/torvalds/linux/blob/ddf21bd8ab984ccaa924f090fc7f515bb6d51414/fs/isofs/dir.c#L28-L29. Feels like we should just match that behaviour?

Reply:
Yeah, but mostly it doesn't, because the filesystem usually has Rock Ridge or Joliet extensions.  I'm not saying the behavior here is wrong, I'm saying it seems more likely to cause c-i bugs when someone doesn't realize that the filenames here are different than the ones they see in `ls`.  In the old Windows days, if you saw a list of uppercase shortnames, you knew you were in 8.3 land.

Reply:
Discussed this OOB.  Let's land as-is for now, and circle back in a followup.